atproto/crypto/keys.go

```diff
@@ -0,0 +1,457 @@
+type PublicKey struct {
+	keyType KeyType
+	pubP256 *ecdsa.PublicKey
+	pubK256 *secp256k1secec.PublicKey
+}
+
+var k256Options = &secp256k1secec.ECDSAOptions{
+	// Used to *verify* digest, not to re-hash
+	Hash: crypto.SHA256,
+	// Use `[R | S]` encoding.
+	Encoding: secp256k1secec.EncodingCompact,
+	// Checking `s <= n/2` to prevent signature mallability is not part of SEC 1, Version 2.0. libsecp256k1 which used to be used by this package, includes the check, so retain behavior compatibility.
+	RejectMalleable: true,
+}
+
+// Creates a secure new cryptographic key from scratch, with the indicated curve type.
+func GeneratePrivateKey(kt KeyType) (*PrivateKey, error) {
+	switch kt {
+	case P256:
+		key, err := ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return nil, fmt.Errorf("P-256/secp256r1 key generation failed: %w", err)
+		}
+		return &PrivateKey{keyType: kt, privP256: key}, nil
+	case K256:
+		key, err := secp256k1secec.GenerateKey()
+		if err != nil {
+			return nil, fmt.Errorf("K-256/secp256k1 key generation failed: %w", err)
+		}
+		return &PrivateKey{keyType: kt, privK256: key}, nil
+	default:
+		return nil, fmt.Errorf("unexpected crypto KeyType")
+	}
+}
+
+// Loads a [PrivateKey] of the indicated curve type from raw bytes, as exported by the [PrivateKey.Bytes()] method.
+//
+// Calling code needs to know the key type ahead of time, and must remove any string encoding (hex encoding, base64, etc) before calling this function.
+func ParsePrivateKeyBytes(data []byte, kt KeyType) (*PrivateKey, error) {
+	switch kt {
+	case P256:
+		// elaborately parse as an ecdh.PrivateKey, then get from that to ecdsa.PrivateKey by encoding/decoding using x509 PKCS8 encoding.
+		// Note that the 'data' bytes format is *not* x509 PKCS8!
+		skEcdh, err := ecdh.P256().NewPrivateKey(data)
+		if err != nil {
+			return nil, fmt.Errorf("invalid P-256/secp256r1 private key: %w", err)
+		}
+		enc, err := x509.MarshalPKCS8PrivateKey(skEcdh)
+		if err != nil {
+			return nil, fmt.Errorf("invalid P-256/secp256r1 private key: %w", err)
+		}
+		sk, err := x509.ParsePKCS8PrivateKey(enc)
+		if err != nil {
+			return nil, fmt.Errorf("invalid P-256/secp256r1 private key: %w", err)
+		}
+		return &PrivateKey{keyType: kt, privP256: sk.(*ecdsa.PrivateKey)}, nil
+	case K256:
+		sk, err := secp256k1secec.NewPrivateKey(data)
+		if err != nil {
+			return nil, fmt.Errorf("invalid K-256/secp256k1 private key: %w", err)
+		}
+		return &PrivateKey{keyType: kt, privK256: sk}, nil
+	default:
+		return nil, fmt.Errorf("unexpected crypto KeyType")
+	}
+}
+
+// Checks if the two private keys are the same. Note that the naive == operator does not work for most equality checks.
+func (k *PrivateKey) Equal(other *PrivateKey) bool {
+	if k.keyType != other.keyType {
+		return false
+	}
+	switch k.keyType {
+	case P256:
+		return k.privP256.Equal(other.privP256)
+	case K256:
+		return k.privK256.Equal(other.privK256)
+	default:
+		panic("unexpected crypto KeyType")
```

Comment:
Do we need to `panic` here? Can we have this function return an error? It might be helpful to return an error for mismatched key types too.